test/extended/machines/machines.go

```diff
@@ -100,6 +116,91 @@ var _ = g.Describe("[sig-cluster-lifecycle][Feature:Machines] Managed cluster sh
 			e2e.Failf("Machine resources missing for nodes: %s", strings.Join(nodeNames.List(), ", "))
 		}
 	})
+
+	g.It("[sig-scheduling][Early] control plane machine set operator should not cause an early rollout", func() {
+		machineClientSet, err := machineclient.NewForConfig(oc.KubeFramework().ClientConfig())
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		pattern := `^([a-zA-Z0-9]+-)+master-\d+$`
+
+		g.By("checking for the openshift machine api operator")
+		skipUnlessMachineAPIOperator(dc, c.CoreV1().Namespaces())
+
+		g.By("ensuring every node is linked to a machine api resource")
+		allControlPlaneMachines, err := machineClientSet.MachineV1beta1().Machines("openshift-machine-api").List(context.Background(), metav1.ListOptions{
+			LabelSelector: masterMachineLabelSelector,
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		regex, err := regexp.Compile(pattern)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		for _, m := range allControlPlaneMachines.Items {
+			matched := regex.MatchString(m.Name)
+			o.Expect(matched).To(o.BeTrue(), fmt.Sprintf("unexpected name of a control machine occured during early stages: %s", m.Name))
+		}
+	})
+
+	g.It("[sig-scheduling][Early] control plane machine set operator should not have any events", func() {
+		ctx := context.Background()
+
+		nodeClient := oc.KubeClient().CoreV1().Nodes()
+		nodeList, err := nodeClient.List(ctx, metav1.ListOptions{})
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		// We want to skip this test on single-node clusters
+		// as the control plane machine set does not get generated.
+		// No other topology should match this condition.
+		if len(nodeList.Items) == 1 {
+			_, isWorker := nodeList.Items[0].Labels["node-role.kubernetes.io/worker"]
+			_, isControlPlane := nodeList.Items[0].Labels["node-role.kubernetes.io/control-plane"]
+
+			if isWorker && isControlPlane {
+				g.Skip("Skipping test due to a cluster being a single node cluster")
+			}
+		}
+
+		configClient, err := configclient.NewForConfig(oc.KubeFramework().ClientConfig())
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		infrastructure, err := configClient.Infrastructures().Get(ctx, "cluster", metav1.GetOptions{})
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		platform := infrastructure.Status.PlatformStatus.Type
+
+		switch platform {
+		case v1.AWSPlatformType,
+			v1.AzurePlatformType,
+			v1.GCPPlatformType,
+			v1.NutanixPlatformType,
+			v1.OpenStackPlatformType:
+			// Continue with the test
+		default:
+			g.Skip(fmt.Sprintf("Skipping test on platform: %s", platform))
+		}
+
+		machineClientSet, err := machineclient.NewForConfig(oc.KubeFramework().ClientConfig())
+		o.Expect(err).ToNot(o.HaveOccurred())
+
+		g.By("checking for the openshift machine api operator")
+		skipUnlessMachineAPIOperator(dc, c.CoreV1().Namespaces())
```

Comment:
Just for my education, do we run machine api operator even on SNO?

Reply:
There are instances of SNO with MAO yes, I believe you can add additional worker machines to SNO using MAPI unless you've disabled it with the capabilities API